### DISCUSSION TITLE
Purge Submissions - Purge User Uploads

### DISCUSSION THREAD

**Discussion by kiwidood** [solspace/craft-freeform#466] [Feature Requests]
Purge Submissions only deletes the database entries but it doesn't delete the user uploaded files.

This can lead to company compliance issues, where personally identifiable information cannot be stored for more than 30 days for example.

It would be good if there was a manual tool to purge user uploads that are not currently linked to any existing submissions in the database.

It would also be good if the Purge Submissions functionality automatically deletes uploads associated to a submission that's deleted. As a start it would good if it was documented that these files do not get deleted automatically.

**Comment by kjmartens** [comment]
Hi @kiwidood,

Freeform should be deleting the uploaded files as well, but it does this in a separate process, every 3 hours. It's possible that it could take a bit longer, depending on when the last check was, etc.

If this is still not happening, can you tell me which version of Freeform you're experiencing this on? 🙂

**Reply by kiwidood** [reply]
Hi,

We're running Craft Pro 4.2.5.1 with Freeform Pro 4.0.4.

From my investigation purging doesn't seem to be doing anything, other than hiding the submissions in the dashboard (due to them being soft deleted). Nothing ever gets removed at current files, or submissions in the database.

**File Upload Purging Issues**

I ran through a bunch of things on our staging site that I thought it could be (permissions, using a symlink on the asset directory etc.) Same problem.

I then ran it locally and the issue was still happening, so I debugged the code.

The only thing that fixes the issue is removing "->skipContent(true)" from line 452 in "vendor/solspace/craft-freeform/packages/plugin/src/Services/SubmissionsService.php". When it goes to get the value of the upload, it never returns anything in line 466 due to the skipContent(true) part.

I assume this is a bug and potentially has been there for a while? We have user-uploads since 2020.

I was testing this using: "php craft freeform/purge/submissions -d 1" and it would always say "Purge 0 assets" until I changed the code.

Now there's still the issue of existing user-uploads that weren't purged in the past. You mentioned a separate process that runs every 3 hours, is this intended to clean up user-uploads that no longer have submissions associated or were you referring to the one that is executed by the command above?

I assume there's a separate process that is meant to delete all user-uploads that aren't in any submissions? If not there could be remaining user-uploads if there were I/O issues during the deleteElement call or if any other issues occurred.

**Submission Purging Issues**

I have come an issue in line 472 of the same file, it's soft deleting the submission element instead of hard deleting them (second parameter 'true' is not set). It appears that this isn't intended as there's no reference to the GC event in your code. I updated the 'dateDeleted' in the elements table for elements of type "Solspace\Freeform\Elements\Submission" and ran the craft GC command. They never get deleted.

Thanks for your help!

**Reply by kjmartens** [reply]
Sorry for the long delay on this @kiwidood. It somehow fell through the cracks. We released an update (4.0.6) which addresses an issue where garbage collection was not working with Freeform, so I believe this should be resolved for you now. If you haven't already, can you try updating and see if the issue is resolved? 🙂
